File: 15.3-sum.ts
/*
 * @lc app=leetcode id=15 lang=typescript
 *
 * [15] 3Sum
 */

// @lc code=start
function threeSum(nums: number[]): number[][] {
  const numToCount = new Map<number, number>();
  for (let i = 0; i < nums.length; i++) {
    numToCount.set(nums[i], (numToCount.get(nums[i]) ?? 0) + 1);
  }

  const answers: number[][] = [];
  const zeroCount = numToCount.get(0);
  if (zeroCount !== undefined && zeroCount >= 3) {
    answers.push([0, 0, 0]);
  }

  const candidateNums = [...numToCount.keys()];
  const positives = candidateNums.filter((n) => n > 0);
  const negatives = candidateNums.filter((n) => n < 0);

  for (let i = 0; i < positives.length; i++) {
    for (let j = 0; j < negatives.length; j++) {
      const num1 = positives[i];
      const num2 = negatives[j];
      const rest = -(num1 + num2);
      if (numToCount.get(rest) === undefined) continue;
      if ((num1 === rest && (numToCount.get(num1) ?? 0) < 2) || (num2 === rest && (numToCount.get(num2) ?? 0) < 2)) {
        continue;
      }
      answers.push([num1, num2, rest].sort((a, b) => a - b));
    }
  }

  return answers;
}
// @lc code=end

File: 15.3-sum.rb
#
# @lc app=leetcode id=15 lang=ruby
#
# [15] 3Sum
#

# @lc code=start
# @param {Integer[]} nums
# @return {Integer[][]}
require 'set'
def three_sum(nums)
  # Cannot solve because of time limit
  hash = Hash.new(0)
  nums.each do |num|
    hash[num] += 1
  end

  res = Set.new
  keys = hash.keys
  keys.each do |key1|
    hash[key1] -= 1

    keys.each do |key2|
      next if hash[key2] == 0

      hash[key2] -= 1

      two_sum = key1 + key2

      if hash[-two_sum] && hash[-two_sum] > 0
        res.add([key1, key2, -two_sum].sort)
      end

      hash[key2] += 1
    end

    hash[key1] += 1
  end

  res.to_a
end
# @lc code=end

p(three_sum([-1, 0, 1, 2, -1, -4]))
